app/controllers/KairosDBProxy.java

```diff
@@ -0,0 +1,349 @@
+     * @return Proxied health check response.
+     */
+    public CompletionStage<Result> healthCheck() {
+        return proxy();
+    }
+
+    /**
+     * Proxied tagNames call.
+     *
+     * @return Proxied tagNames response.
+     */
+    public CompletionStage<Result> tagNames() {
+        return proxy();
+    }
+
+    /**
+     * Proxied tagValues call.
+     *
+     * @return Proxied tagValues response.
+     */
+    public CompletionStage<Result> tagValues() {
+        return proxy();
+    }
+
+    /**
+     * Proxied queryTags call.
+     *
+     * @return Proxied queryTags response.
+     */
+    public CompletionStage<Result> queryTags() {
+        return proxy();
+    }
+
+    /**
+     * Proxied queryMetrics call.
+     *
+     * @return Proxied queryMetrics response.
+     */
+    public CompletionStage<Result> queryMetrics() {
+        return proxy();
+    }
+
+    /**
+     * Proxied version call.
+     *
+     * @return Proxied version response.
+     */
+    public CompletionStage<Result> version() {
+        return proxy();
+    }
+
+    /**
+     * Caching metricNames call.
+     *
+     * @param containing simple string match filter for metric names
+     * @return Cached metric names, filtered by the query string.
+     */
+    public CompletionStage<Result> metricNames(final String containing) {
+        final List<String> metrics = _cache.getIfPresent(METRICS_KEY);
+
+        final CompletionStage<List<String>> response;
+        if (metrics != null) {
+            response = CompletableFuture.completedFuture(metrics);
+        } else {
+            // Refresh the cache
+            final CompletionStage<List<String>> queryResponse = _kairosDbClient.queryMetricNames()
+                    .thenApply(KairosMetricNamesQueryResponse::getResults)
+                    .thenApply(list -> {
+                        _cache.put(METRICS_KEY, list);
+                        _metricsList.set(list);
+                        return list;
+                    });
+
+            final List<String> metricsList = _metricsList.get();
+            if (metricsList != null) {
+                response = CompletableFuture.completedFuture(metricsList);
+            } else {
+                response = queryResponse;
+            }
+        }
+
+        return response
+                .thenApply(list -> filterMetricNames(list, containing))
+                .thenApply(list -> new KairosMetricNamesQueryResponse.Builder().setResults(list).build())
+                .<JsonNode>thenApply(_mapper::valueToTree)
+                .thenApply(Results::ok);
+    }
+
+    private static List<String> filterMetricNames(final List<String> metricNames, final String containing) {
+        if (containing == null || containing.isEmpty()) {
+            return metricNames;
+        }
+
+        final String lowerContaining = containing.toLowerCase(Locale.ENGLISH);
+        final Predicate<String> containsPredicate = s -> s.toLowerCase().contains(lowerContaining);
```

Comment:
extract these into statics?

Reply:
Unfortunately this one can't really be extracted to a static because it closes over lowerContaining.  I moved the other predicate, though.